lib/schedulers.go

```diff
@@ -0,0 +1,331 @@
+	sort.SliceStable(trackedSteps, func(a, b int) bool {
+		switch {
+		case trackedSteps[a].TimeOffset < trackedSteps[b].TimeOffset:
+			return true
+		case trackedSteps[a].TimeOffset == trackedSteps[b].TimeOffset:
+			return trackedSteps[a].configID < trackedSteps[b].configID
+		default:
+			return false
+		}
+	})
+
+	// Go through all of the sorted steps from all of the schedulers, and
+	// build a new list of execution steps that consolidates all of their
+	// requirements. If multiple schedulers have an execution step at exactly
+	// the same time offset, they will be combined into a single new execution
+	// step with the sum of the values from the previous ones.
+	currentTimeOffset := time.Duration(0)
+	currentPlannedVUs := make([]uint64, len(scs))
+	currentMaxUnplannedVUs := make([]uint64, len(scs))
+	sum := func(data []uint64) (result uint64) { // sigh...
+		for _, val := range data {
+			result += val
+		}
+		return result
+	}
+	consolidatedSteps := []ExecutionStep{}
+	addCurrentStepIfDifferent := func() {
+		newPlannedVUs := sum(currentPlannedVUs)
+		newMaxUnplannedVUs := sum(currentMaxUnplannedVUs)
+		stepsLen := len(consolidatedSteps)
+		if stepsLen == 0 ||
+			consolidatedSteps[stepsLen-1].PlannedVUs != newPlannedVUs ||
+			consolidatedSteps[stepsLen-1].MaxUnplannedVUs != newMaxUnplannedVUs {
+
+			consolidatedSteps = append(consolidatedSteps, ExecutionStep{
+				TimeOffset:      currentTimeOffset,
+				PlannedVUs:      newPlannedVUs,
+				MaxUnplannedVUs: newMaxUnplannedVUs,
+			})
+		}
+	}
+	for _, step := range trackedSteps {
+		// If the time offset is different, create a new step with the current values
+		if step.TimeOffset != currentTimeOffset {
+			addCurrentStepIfDifferent()
+			currentTimeOffset = step.TimeOffset
+		}
+		currentPlannedVUs[step.configID] = step.PlannedVUs
+		currentMaxUnplannedVUs[step.configID] = step.MaxUnplannedVUs
+	}
+	addCurrentStepIfDifferent() // Add the last step
+	return consolidatedSteps
+}
+
+// GetParsedSchedulerConfig returns a struct instance corresponding to the supplied
+// config type. It will be fully initialized - with both the default values of
+// the type, as well as with whatever the user had specified in the JSON
+func GetParsedSchedulerConfig(name, configType string, rawJSON []byte) (result SchedulerConfig, err error) {
+	schedulerConfigTypesMutex.Lock()
+	defer schedulerConfigTypesMutex.Unlock()
+
+	constructor, exists := schedulerConfigConstructors[configType]
+	if !exists {
+		return nil, fmt.Errorf("unknown execution scheduler type '%s'", configType)
+	}
+	return constructor(name, rawJSON)
+}
+
+type protoSchedulerConfig struct {
+	configType string
+	rawJSON    json.RawMessage
+}
+
+// UnmarshalJSON just reads unmarshals the base config (to get the type), but it also
+// stores the unprocessed JSON so we can parse the full config in the next step
+func (pc *protoSchedulerConfig) UnmarshalJSON(b []byte) error {
+	var tmp struct {
+		ConfigType string `json:"type"`
+	}
```

Comment:
Not certain if we can't just have the `protoSchedulerConfig` to have it's `configType` be `ConfigType` or just `Type` and use it instead.

Reply:
Unfortunately not :grinning: I originally tried it that way and got endless recursion... Basically, we need the custom `UnmarshalJSON()` method so we can save the raw JSON data for post-processing, but we can't call `json.Unmarshal()` inside of it on our own `protoSchedulerConfig` type, since it will just call our `UnmarshalJSON()` method again.